apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1360,8 +1395,221 @@ private void StartPush()
+            {
+                _concurrentPushCount -= 1;
+                _pushInFlight -= ackIds?.Count ?? 0 + nackIds?.Count ?? 0 + extendIds?.Count ?? 0;
+
+                bool hasAckIds = ackIds?.Count > 0;
+                bool hasNackIds = nackIds?.Count > 0;
+                bool hasExtendIds = extendIds?.Count > 0;
+                // In one call, we 'll have one of acks or nacks or extends, not all.
+                var ids = hasAckIds ? ackIds
+                    : hasNackIds ? nackIds : extendIds.Select(j => j.Id);
+
+                if (writeTask.IsFaulted)
+                {
+                    // There may be few successes and few failures or all failures.
+                    var rpcException = writeTask.Exception.As<RpcException>();
+                    if (rpcException != null)
+                    {
+                        var ackError = AckError.ForRpcException(rpcException, ids);
+                        // Only temporary failures should be retried. Mark them as retryable if not already.
+                        UpdateRetryableIds(ids, ackError);
+
+                        if (hasAckIds)
+                        {
+                            // ack = true implies acks.
+                            RetryAcksAndExtends(true, ackError, Enumerable.Empty<TimedId>());
+                        }
+
+                        if (hasNackIds)
+                        {
+                            // ack = false implies nacks.
+                            RetryAcksAndExtends(false, ackError, Enumerable.Empty<TimedId>());
+                        }
+
+                        if (hasExtendIds)
+                        {
+                            // ack = null implies extends.
+                            RetryAcksAndExtends(null, ackError, extendIds);
+                        }
+                    }
+                    else
+                    {
+                        // Throw it.
+                        throw writeTask.Exception.FlattenIfPossible();
+                    }
+                }
+                else
+                {
+                    // Everything succeeded. Update _retryableIds dictionary.
+                    UpdateRetryableIds(ids, default);
+                }
+
+                // Perform push so that other messages can be processed.
+                StartPush();
+
+                // This method schedules the retry of ackIds/nackIds/extendIds with temporary errors and
+                // handles the permanent errors. Permanent errors are propagated to the caller for Acknowledge RPC,
+                // while permanent errors are only logged for ModifyAckDeadline RPC as ModifyAckDeadline call is not user initiated.
+                void RetryAcksAndExtends(bool? ack, AckError ackError, IEnumerable<TimedId> extendIds)
+                {
+                    // ack = true implies acks. ack = false implies nacks. ack = null implies extends, in which case extendIds should be non-empty.
```

Comment:
I'd just allow extendIds to be null for non-ack/nack. That's simpler to check for than using `Any` etc - it means it's easier to call, too. I *think* you can just call `RetryAcksAndExtends(hasAckIds, ackError, extendIds)`, if you use `extendIds` being non-null as a signal that it's definitely extend rather than ack/nack. (Or you may find you don't really need `hasAckIds` and `hasNackIds`, and can just use `RetryAcksAndExtends(ackIds is not null, ackError, extendIds)`

Reply:
Or we may find that this really should be two methods after all... anything where one parameter is only relevant half the time feels odd. I might try refactoring it as another commit myself, if you'd like me to. (That could potentially come later.)